File: src/main/ceylon/com/diabolicallabs/process/manager/ceylon/kieprocess/service/KnowledgeService.java
package com.diabolicallabs.process.manager.ceylon.kieprocess.service;

import com.redhat.ceylon.compiler.java.metadata.Ceylon;
import com.redhat.ceylon.compiler.java.metadata.TypeInfo;
import com.redhat.ceylon.compiler.java.metadata.TypeParameter;
import com.redhat.ceylon.compiler.java.metadata.TypeParameters;
import com.redhat.ceylon.compiler.java.metadata.Variance;
import com.redhat.ceylon.compiler.java.metadata.Ignore;
import com.redhat.ceylon.compiler.java.metadata.Name;
import com.redhat.ceylon.compiler.java.runtime.model.TypeDescriptor;
import com.redhat.ceylon.compiler.java.runtime.model.ReifiedType;
import ceylon.language.Callable;
import ceylon.language.DocAnnotation$annotation$;
import io.vertx.core.json.JsonArray;
import io.vertx.ceylon.core.Vertx;
import io.vertx.core.AsyncResult;
import io.vertx.core.Handler;

@Ceylon(major = 8)
public class KnowledgeService implements ReifiedType {

  @Ignore
  public static final io.vertx.lang.ceylon.ConverterFactory<com.diabolicallabs.process.manager.service.KnowledgeService, KnowledgeService> TO_CEYLON = new io.vertx.lang.ceylon.ConverterFactory<com.diabolicallabs.process.manager.service.KnowledgeService, KnowledgeService>() {
    public io.vertx.lang.ceylon.Converter<com.diabolicallabs.process.manager.service.KnowledgeService, KnowledgeService> converter(final TypeDescriptor... descriptors) {
      return new io.vertx.lang.ceylon.Converter<com.diabolicallabs.process.manager.service.KnowledgeService, KnowledgeService>() {
        public KnowledgeService convert(com.diabolicallabs.process.manager.service.KnowledgeService src) {
          return new KnowledgeService(src);
        }
      };
    }
  };

  @Ignore
  public static final io.vertx.lang.ceylon.Converter<KnowledgeService, com.diabolicallabs.process.manager.service.KnowledgeService> TO_JAVA = new io.vertx.lang.ceylon.Converter<KnowledgeService, com.diabolicallabs.process.manager.service.KnowledgeService>() {
    public com.diabolicallabs.process.manager.service.KnowledgeService convert(KnowledgeService src) {
      return src.delegate;
    }
  };

  @Ignore public static final TypeDescriptor $TypeDescriptor$ = TypeDescriptor.klass(KnowledgeService.class);
  @Ignore private final com.diabolicallabs.process.manager.service.KnowledgeService delegate;

  public KnowledgeService(com.diabolicallabs.process.manager.service.KnowledgeService delegate) {
    this.delegate = delegate;
  }

  @Ignore 
  public TypeDescriptor $getType$() {
    return $TypeDescriptor$;
  }

  @Ignore
  public Object getDelegate() {
    return delegate;
  }

  @TypeInfo("com.diabolicallabs.process.manager.ceylon.kieprocess.service::KnowledgeService")
  public KnowledgeService addClassPathResource(
    final @TypeInfo("ceylon.language::String") @Name("resourceName")  ceylon.language.String resourceName, 
    final @TypeInfo("ceylon.language::Anything(ceylon.language::Throwable?)") @Name("handler")  Callable<?> handler) {
    java.lang.String arg_0 = io.vertx.lang.ceylon.ToJava.String.safeConvert(resourceName);
    io.vertx.core.Handler<io.vertx.core.AsyncResult<java.lang.Void>> arg_1 = handler == null ? null : new io.vertx.lang.ceylon.CallableAsyncResultHandler<java.lang.Void>(handler) {
      public Object toCeylon(java.lang.Void event) {
        return null;
      }
    };
    KnowledgeService ret = com.diabolicallabs.process.manager.ceylon.kieprocess.service.KnowledgeService.TO_CEYLON.converter().safeConvert(delegate.addClassPathResource(arg_0, arg_1));
    return this;
  }

  @TypeInfo("com.diabolicallabs.process.manager.ceylon.kieprocess.service::KnowledgeService")
  public KnowledgeService addFileResource(
    final @TypeInfo("ceylon.language::String") @Name("fileName")  ceylon.language.String fileName, 
    final @TypeInfo("ceylon.language::Anything(ceylon.language::Throwable?)") @Name("handler")  Callable<?> handler) {
    java.lang.String arg_0 = io.vertx.lang.ceylon.ToJava.String.safeConvert(fileName);
    io.vertx.core.Handler<io.vertx.core.AsyncResult<java.lang.Void>> arg_1 = handler == null ? null : new io.vertx.lang.ceylon.CallableAsyncResultHandler<java.lang.Void>(handler) {
      public Object toCeylon(java.lang.Void event) {
        return null;
      }
    };
    KnowledgeService ret = com.diabolicallabs.process.manager.ceylon.kieprocess.service.KnowledgeService.TO_CEYLON.converter().safeConvert(delegate.addFileResource(arg_0, arg_1));
    return this;
  }

  @TypeInfo("com.diabolicallabs.process.manager.ceylon.kieprocess.service::KnowledgeService")
  public KnowledgeService processDefinitions(
    final @TypeInfo("ceylon.language::Anything(ceylon.language::Throwable|ceylon.json::Array)") @Name("handler")  Callable<?> handler) {
    io.vertx.core.Handler<io.vertx.core.AsyncResult<io.vertx.core.json.JsonArray>> arg_0 = handler == null ? null : new io.vertx.lang.ceylon.CallableAsyncResultHandler<io.vertx.core.json.JsonArray>(handler) {
      public Object toCeylon(io.vertx.core.json.JsonArray event) {
        return io.vertx.lang.ceylon.ToCeylon.JsonArray.safeConvert(event);
      }
    };
    KnowledgeService ret = com.diabolicallabs.process.manager.ceylon.kieprocess.service.KnowledgeService.TO_CEYLON.converter().safeConvert(delegate.processDefinitions(arg_0));
    return this;
  }

  @TypeInfo("com.diabolicallabs.process.manager.ceylon.kieprocess.service::KnowledgeService")
  public KnowledgeService getProcessService(
    final @TypeInfo("ceylon.language::Anything(ceylon.language::Throwable|com.diabolicallabs.process.manager.ceylon.kieprocess.service::ProcessService)") @Name("handler")  Callable<?> handler) {
    io.vertx.core.Handler<io.vertx.core.AsyncResult<com.diabolicallabs.process.manager.service.ProcessService>> arg_0 = handler == null ? null : new io.vertx.lang.ceylon.CallableAsyncResultHandler<com.diabolicallabs.process.manager.service.ProcessService>(handler) {
      public Object toCeylon(com.diabolicallabs.process.manager.service.ProcessService event) {
        return com.diabolicallabs.process.manager.ceylon.kieprocess.service.ProcessService.TO_CEYLON.converter().safeConvert(event);
      }
    };
    KnowledgeService ret = com.diabolicallabs.process.manager.ceylon.kieprocess.service.KnowledgeService.TO_CEYLON.converter().safeConvert(delegate.getProcessService(arg_0));
    return this;
  }

  @TypeInfo("com.diabolicallabs.process.manager.ceylon.kieprocess.service::KnowledgeService")
  public KnowledgeService getRuleService(
    final @TypeInfo("ceylon.language::Anything(ceylon.language::Throwable|com.diabolicallabs.process.manager.ceylon.kieprocess.service::RuleService)") @Name("handler")  Callable<?> handler) {
    io.vertx.core.Handler<io.vertx.core.AsyncResult<com.diabolicallabs.process.manager.service.RuleService>> arg_0 = handler == null ? null : new io.vertx.lang.ceylon.CallableAsyncResultHandler<com.diabolicallabs.process.manager.service.RuleService>(handler) {
      public Object toCeylon(com.diabolicallabs.process.manager.service.RuleService event) {
        return com.diabolicallabs.process.manager.ceylon.kieprocess.service.RuleService.TO_CEYLON.converter().safeConvert(event);
      }
    };
    KnowledgeService ret = com.diabolicallabs.process.manager.ceylon.kieprocess.service.KnowledgeService.TO_CEYLON.converter().safeConvert(delegate.getRuleService(arg_0));
    return this;
  }

  @TypeInfo("com.diabolicallabs.process.manager.ceylon.kieprocess.service::KnowledgeService")
  public KnowledgeService getTaskService(
    final @TypeInfo("ceylon.language::Anything(ceylon.language::Throwable|com.diabolicallabs.process.manager.ceylon.kieprocess.service::TaskService)") @Name("handler")  Callable<?> handler) {
    io.vertx.core.Handler<io.vertx.core.AsyncResult<com.diabolicallabs.process.manager.service.TaskService>> arg_0 = handler == null ? null : new io.vertx.lang.ceylon.CallableAsyncResultHandler<com.diabolicallabs.process.manager.service.TaskService>(handler) {
      public Object toCeylon(com.diabolicallabs.process.manager.service.TaskService event) {
        return com.diabolicallabs.process.manager.ceylon.kieprocess.service.TaskService.TO_CEYLON.converter().safeConvert(event);
      }
    };
    KnowledgeService ret = com.diabolicallabs.process.manager.ceylon.kieprocess.service.KnowledgeService.TO_CEYLON.converter().safeConvert(delegate.getTaskService(arg_0));
    return this;
  }

  @TypeInfo("com.diabolicallabs.process.manager.ceylon.kieprocess.service::KnowledgeService")
  public KnowledgeService getTaskServiceAddress(
    final @TypeInfo("ceylon.language::Anything(ceylon.language::Throwable|ceylon.language::String)") @Name("handler")  Callable<?> handler) {
    io.vertx.core.Handler<io.vertx.core.AsyncResult<java.lang.String>> arg_0 = handler == null ? null : new io.vertx.lang.ceylon.CallableAsyncResultHandler<java.lang.String>(handler) {
      public Object toCeylon(java.lang.String event) {
        return io.vertx.lang.ceylon.ToCeylon.String.safeConvert(event);
      }
    };
    KnowledgeService ret = com.diabolicallabs.process.manager.ceylon.kieprocess.service.KnowledgeService.TO_CEYLON.converter().safeConvert(delegate.getTaskServiceAddress(arg_0));
    return this;
  }

  @TypeInfo("ceylon.language::Anything")
  public void close() {
    delegate.close();
  }

}

File: src/main/resources/kieprocess-js/knowledge_service.js
/*
 * Copyright 2014 Red Hat, Inc.
 *
 * Red Hat licenses this file to you under the Apache License, version 2.0
 * (the "License"); you may not use this file except in compliance with the
 * License.  You may obtain a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

/** @module kieprocess-js/knowledge_service */
var utils = require('vertx-js/util/utils');
var RuleService = require('kieprocess-js/rule_service');
var ProcessService = require('kieprocess-js/process_service');
var Vertx = require('vertx-js/vertx');
var TaskService = require('kieprocess-js/task_service');

var io = Packages.io;
var JsonObject = io.vertx.core.json.JsonObject;
var JKnowledgeService = com.diabolicallabs.process.manager.service.KnowledgeService;

/**
 @class
*/
var KnowledgeService = function(j_val) {

  var j_knowledgeService = j_val;
  var that = this;

  /**

   @public
   @param resourceName {string} 
   @param handler {function} 
   @return {KnowledgeService}
   */
  this.addClassPathResource = function(resourceName, handler) {
    var __args = arguments;
    if (__args.length === 2 && typeof __args[0] === 'string' && typeof __args[1] === 'function') {
      j_knowledgeService["addClassPathResource(java.lang.String,io.vertx.core.Handler)"](resourceName, function(ar) {
      if (ar.succeeded()) {
        handler(null, null);
      } else {
        handler(null, ar.cause());
      }
    });
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public
   @param fileName {string} 
   @param handler {function} 
   @return {KnowledgeService}
   */
  this.addFileResource = function(fileName, handler) {
    var __args = arguments;
    if (__args.length === 2 && typeof __args[0] === 'string' && typeof __args[1] === 'function') {
      j_knowledgeService["addFileResource(java.lang.String,io.vertx.core.Handler)"](fileName, function(ar) {
      if (ar.succeeded()) {
        handler(null, null);
      } else {
        handler(null, ar.cause());
      }
    });
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public
   @param handler {function} 
   @return {KnowledgeService}
   */
  this.processDefinitions = function(handler) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] === 'function') {
      j_knowledgeService["processDefinitions(io.vertx.core.Handler)"](function(ar) {
      if (ar.succeeded()) {
        handler(utils.convReturnJson(ar.result()), null);
      } else {
        handler(null, ar.cause());
      }
    });
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public
   @param handler {function} 
   @return {KnowledgeService}
   */
  this.getProcessService = function(handler) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] === 'function') {
      j_knowledgeService["getProcessService(io.vertx.core.Handler)"](function(ar) {
      if (ar.succeeded()) {
        handler(utils.convReturnVertxGen(ar.result(), ProcessService), null);
      } else {
        handler(null, ar.cause());
      }
    });
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public
   @param handler {function} 
   @return {KnowledgeService}
   */
  this.getRuleService = function(handler) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] === 'function') {
      j_knowledgeService["getRuleService(io.vertx.core.Handler)"](function(ar) {
      if (ar.succeeded()) {
        handler(utils.convReturnVertxGen(ar.result(), RuleService), null);
      } else {
        handler(null, ar.cause());
      }
    });
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public
   @param handler {function} 
   @return {KnowledgeService}
   */
  this.getTaskService = function(handler) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] === 'function') {
      j_knowledgeService["getTaskService(io.vertx.core.Handler)"](function(ar) {
      if (ar.succeeded()) {
        handler(utils.convReturnVertxGen(ar.result(), TaskService), null);
      } else {
        handler(null, ar.cause());
      }
    });
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public
   @param handler {function} 
   @return {KnowledgeService}
   */
  this.getTaskServiceAddress = function(handler) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] === 'function') {
      j_knowledgeService["getTaskServiceAddress(io.vertx.core.Handler)"](function(ar) {
      if (ar.succeeded()) {
        handler(ar.result(), null);
      } else {
        handler(null, ar.cause());
      }
    });
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public

   */
  this.close = function() {
    var __args = arguments;
    if (__args.length === 0) {
      j_knowledgeService["close()"]();
    } else throw new TypeError('function invoked with invalid arguments');
  };

  // A reference to the underlying Java delegate
  // NOTE! This is an internal API and must not be used in user code.
  // If you rely on this property your code is likely to break if we change it / remove it without warning.
  this._jdel = j_knowledgeService;
};

/**

 @memberof module:kieprocess-js/knowledge_service
 @param vertx {Vertx} 
 @param address {string} 
 @return {KnowledgeService}
 */
KnowledgeService.createProxy = function(vertx, address) {
  var __args = arguments;
  if (__args.length === 2 && typeof __args[0] === 'object' && __args[0]._jdel && typeof __args[1] === 'string') {
    return utils.convReturnVertxGen(JKnowledgeService["createProxy(io.vertx.core.Vertx,java.lang.String)"](vertx._jdel, address), KnowledgeService);
  } else throw new TypeError('function invoked with invalid arguments');
};

// We export the Constructor function
module.exports = KnowledgeService;